jetty-quic-quiche-foreign fails on Windows due to eager resolution of Unix-only quiche_conn_set_keylog_fd

**Jetty version(s)**

12.1.9

Also reproduced with 12.1.8.

**Jetty Environment**

core

**HTTP version**

HTTP/3

**Java version/vendor** `(use: java -version)`

Graal JDK 25

**OS type/version**

Windows x86_64

**Description**

`jetty-quic-quiche-foreign` fails to initialize on Windows because the generated foreign binding eagerly resolves `quiche_conn_set_keylog_fd` during static initialization.

The Windows `quiche.dll` from:

org.mortbay.jetty.quiche:jetty-quiche-native:0.29.1
does not export quiche_conn_set_keylog_fd.

dumpbin /exports quiche.dll shows:

quiche_conn_set_keylog_path

but does not show:

quiche_conn_set_keylog_fd

This seems expected on Windows because Cloudflare quiche documents the *_fd logging functions as Unix-only. However, Jetty's generated foreign binding resolves these symbols eagerly in quiche_h$DowncallHandles.<clinit>, so the whole foreign binding fails before HTTP/3 can start.

How to reproduce

Use Windows x86_64 with:

<dependency>
  <groupId>org.eclipse.jetty.quic</groupId>
  <artifactId>jetty-quic-quiche-foreign</artifactId>
  <version>12.1.9</version>
</dependency>

<dependency>
  <groupId>org.mortbay.jetty.quiche</groupId>
  <artifactId>jetty-quiche-native</artifactId>
  <version>0.29.1</version>
</dependency>

<dependency>
  <groupId>org.eclipse.jetty.quic</groupId>
  <artifactId>jetty-quic-quiche-server</artifactId>
  <version>12.1.9</version>
</dependency>

<dependency>
  <groupId>org.eclipse.jetty.http3</groupId>
  <artifactId>jetty-http3-server</artifactId>
  <version>12.1.9</version>
</dependency>

JNA is not on the classpath.

ServiceLoader successfully finds the foreign provider:

Quiche binding provider = org.eclipse.jetty.quic.quiche.foreign.ForeignQuicheBinding
Priority = 100

But initialization fails:

Initialize failure = java.lang.AssertionError: should not reach here
java.lang.AssertionError: should not reach here
    at org.eclipse.jetty.quic.quiche.foreign.quiche_h.quiche_version(quiche_h.java:1016)
    at org.eclipse.jetty.quic.quiche.foreign.quiche_h.initialize(quiche_h.java:38)
    at org.eclipse.jetty.quic.quiche.foreign.ForeignQuicheBinding.initialize(ForeignQuicheBinding.java:35)

Caused by: java.lang.UnsatisfiedLinkError: unresolved symbol: quiche_conn_set_keylog_fd
    at org.eclipse.jetty.quic.quiche.foreign.NativeHelper.lambda$downcallHandle$0(NativeHelper.java:113)
    at java.base/java.util.Optional.orElseThrow(Optional.java:403)
    at org.eclipse.jetty.quic.quiche.foreign.NativeHelper.downcallHandle(NativeHelper.java:113)
    at org.eclipse.jetty.quic.quiche.foreign.quiche_h$DowncallHandles.<clinit>(quiche_h.java:403)
    at org.eclipse.jetty.quic.quiche.foreign.quiche_h.quiche_version(quiche_h.java:1012)

I verified the extracted Windows DLL with:

dumpbin /exports quiche.dll | findstr /i keylog

It exports:

quiche_conn_set_keylog_path

but not:

quiche_conn_set_keylog_fd

Workaround tested

I locally overrode org.eclipse.jetty.quic.quiche.foreign.quiche_h and changed only these two MethodHandles to be unavailable instead of eagerly resolved:

private static final MethodHandle quiche_conn_set_keylog_fd = null;
private static final MethodHandle quiche_conn_set_qlog_fd = null;

After that, ForeignQuicheBinding.initialize() succeeds:

Initialize failure = null

Expected behavior

jetty-quic-quiche-foreign should initialize on Windows.

Unix-only symbols such as:

quiche_conn_set_keylog_fd
quiche_conn_set_qlog_fd

should not be eagerly required on Windows. They should probably be optional/lazy, or the generated binding should avoid resolving Unix-only symbols on Windows.

Actual behavior

The foreign binding eagerly resolves quiche_conn_set_keylog_fd during static initialization, causing ForeignQuicheBinding.initialize() to fail on Windows before HTTP/3 startup.

--- a/jetty-core/jetty-quic/jetty-quic-quiche/jetty-quic-quiche-foreign/src/main/java/org/eclipse/jetty/quic/quiche/foreign/quiche_h.java
+++ b/jetty-core/jetty-quic/jetty-quic-quiche/jetty-quic-quiche-foreign/src/main/java/org/eclipse/jetty/quic/quiche/foreign/quiche_h.java
@@ -400,26 +400,12 @@ public class quiche_h
                 C_POINTER,
                 C_POINTER
             ));
-        private static final MethodHandle quiche_conn_set_keylog_fd = NativeHelper.downcallHandle(
-            "quiche_conn_set_keylog_fd",
-            FunctionDescriptor.ofVoid(
-                C_POINTER,
-                C_INT
-            ));
         private static final MethodHandle quiche_conn_set_qlog_path = NativeHelper.downcallHandle(
             "quiche_conn_set_qlog_path",
             FunctionDescriptor.of(
                 C_BOOL,
                 C_POINTER,
                 C_POINTER,
-                C_POINTER,
-                C_POINTER
-            ));
-        private static final MethodHandle quiche_conn_set_qlog_fd = NativeHelper.downcallHandle(
-            "quiche_conn_set_qlog_fd",
-            FunctionDescriptor.ofVoid(
-                C_POINTER,
-                C_INT,
                 C_POINTER,
                 C_POINTER
             ));
@@ -1557,35 +1543,11 @@ public class quiche_h
         }
     }
 
-    public static void quiche_conn_set_keylog_fd(MemorySegment conn, int fd)
-    {
-        try
-        {
-            DowncallHandles.quiche_conn_set_keylog_fd.invokeExact(conn, fd);
-        }
-        catch (Throwable x)
-        {
-            throw new AssertionError("should not reach here", x);
-        }
-    }
-
     public static boolean quiche_conn_set_qlog_path(MemorySegment conn, MemorySegment path, MemorySegment log_title, MemorySegment log_desc)
     {
         try
         {
             return (byte)DowncallHandles.quiche_conn_set_qlog_path.invokeExact(conn, path, log_title, log_desc) != 0;
-        }
-        catch (Throwable x)
-        {
-            throw new AssertionError("should not reach here", x);
-        }
-    }
-
-    public static void quiche_conn_set_qlog_fd(MemorySegment conn, int fd, MemorySegment log_title, MemorySegment log_desc)
-    {
-        try
-        {
-            DowncallHandles.quiche_conn_set_qlog_fd.invokeExact(conn, fd, log_title, log_desc);
         }
         catch (Throwable x)
         {

